File: cameraview/src/main/java/com/base/cameraview/engine/lock/BaseLock.java
package com.base.cameraview.engine.lock;

import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import com.base.cameraview.engine.action.ActionHolder;
import com.base.cameraview.engine.action.BaseAction;


@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
public abstract class BaseLock extends BaseAction {

    @Override
    public final void onStart(@NonNull ActionHolder holder) {
        super.onStart(holder);
        boolean isSkipped = checkShouldSkip(holder);
        boolean isSupported = checkIsSupported(holder);
        if (isSupported && !isSkipped) {
            onStarted(holder);
        } else {
            setState(STATE_COMPLETED);
        }
    }

    protected abstract void onStarted(@NonNull ActionHolder holder);

    protected abstract boolean checkShouldSkip(@NonNull ActionHolder holder);

    protected abstract boolean checkIsSupported(@NonNull ActionHolder holder);
}

File: cameraview/src/main/java/com/base/cameraview/engine/lock/BaseLock.kt
package com.base.cameraview.engine.lock

import com.base.cameraview.engine.action.Action
import com.base.cameraview.engine.action.ActionHolder
import com.base.cameraview.engine.action.BaseAction

abstract class BaseLock : BaseAction() {
    override fun onStart(holder: ActionHolder) {
        super.onStart(holder)
        val isSkipped = checkShouldSkip(holder)
        val isSupported = checkIsSupported(holder)
        if (isSupported && !isSkipped) {
            onStarted(holder)
        } else {
            state = Action.Companion.STATE_COMPLETED
        }
    }

    protected abstract fun onStarted(holder: ActionHolder)

    protected abstract fun checkShouldSkip(holder: ActionHolder): Boolean

    protected abstract fun checkIsSupported(holder: ActionHolder): Boolean
}
